optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/spotroster/SpotRosterViewportFactory.java

```diff
@@ -0,0 +1,151 @@
+
+    @Inject
+    private TenantStore tenantStore;
+
+    @Inject
+    private CssGridLinesFactory cssGridLinesFactory;
+
+    @Inject
+    private TicksFactory<LocalDateTime> ticksFactory;
+
+    private Map<Spot, Map<ShiftView, TimeSlot>> spotRosterModel;
+
+    private LinearScale<LocalDateTime> scale;
+
+    public SpotRosterViewport getViewport(final SpotRosterView spotRosterView) {
+
+        shiftBlobViewPool.init(1500L, shiftBlobViewInstances::get); //FIXME: Make maxSize variable
+
+        spotRosterModel = buildSpotRosterModel(spotRosterView);
+
+        scale = new Positive2HoursScale(spotRosterView.getStartDate().atTime(0, 0),
+                                        spotRosterView.getEndDate().atTime(0, 0));
+
+        return new SpotRosterViewport(tenantStore.getCurrentTenantId(),
+                                      shiftBlobViewPool::get,
+                                      scale,
+                                      cssGridLinesFactory.newWithSteps(2L, 12L),
+                                      ticksFactory.newTicks(scale, 2L, 12L),
+                                      buildLanes(spotRosterView)
+        );
+    }
+
+    private Map<Spot, Map<ShiftView, TimeSlot>> buildSpotRosterModel(final SpotRosterView spotRosterView) {
+
+        final Map<Long, Spot> spotsById = indexById(spotRosterView.getSpotList());
+        final Map<Long, TimeSlot> timeSlotsById = indexById(spotRosterView.getTimeSlotList());
+
+        return spotRosterView.getTimeSlotIdToSpotIdToShiftViewListMap().values().stream()
+                .flatMap(s -> s.values().stream())
+                .flatMap(Collection::stream)
+                .collect(groupingBy(shiftView -> spotsById.get(shiftView.getSpotId()),
+                                    toMap(identity(), shiftView -> timeSlotsById.get(shiftView.getTimeSlotId()))));
+    }
+
+    private List<Lane<LocalDateTime>> buildLanes(final SpotRosterView spotRosterView) {
+
+        final Map<Long, Employee> employeesById = indexById(spotRosterView.getEmployeeList());
+
+        return spotRosterModel
+                .entrySet()
+                .stream()
+                .map(e -> new SpotLane(e.getKey(), buildSubLanes(e.getKey(), e.getValue(), employeesById)))
+                .collect(toList());
+    }
+
+    private List<SubLane<LocalDateTime>> buildSubLanes(final Spot spot,
+                                                       final Map<ShiftView, TimeSlot> timeSlotsByShift,
+                                                       final Map<Long, Employee> employeesById) {
+
+        //FIXME: Handle overlapping blobs and discover why some TimeSlots are null
+
+        if (timeSlotsByShift.isEmpty()) {
+            return new ArrayList<>(singletonList(new SubLane<>()));
+        }
+
+        final List<Blob<LocalDateTime>> blobs = timeSlotsByShift.entrySet()
+                .stream()
+                .filter(e -> e.getValue() != null) //FIXME: Why are there null Time Slots?
```

Comment:
Assuming you build the maps the same way as I did, the reason for the nulls would be some spots do not have shifts within the time slots (https://github.com/kiegroup/optashift-employee-rostering/blob/2dc8bdcb953e6d8b96ea4357fc107cec036ed143/optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/spot/SpotDataFetchable.java#L78).